src/random.h

```diff
@@ -135,50 +135,68 @@ void RandAddPeriodic() noexcept;
+template<typename T>
+class RandomMixin;
+
+/** A concept for RandomMixin-based random number generators. */
+template<typename T>
+concept RandomNumberGenerator = requires(T& rng, Span<std::byte> s) {
+    // A random number generator must provide rand64().
+    { rng.rand64() } noexcept -> std::same_as<uint64_t>;
+    // A random number generator must provide randfill(Span<std::byte>).
+    { rng.fillrand(s) } noexcept;
+    // A random number generator must derive from RandomMixin, which adds other rand* functions.
+    requires std::derived_from<std::remove_reference_t<T>, RandomMixin<std::remove_reference_t<T>>>;
+};
+
+/** Mixin class that provides helper randomness functions.
  *
- * This class is not thread-safe.
+ * Intended to be used through CRTP: https://en.cppreference.com/w/cpp/language/crtp.
+ * An RNG class FunkyRNG would derive publicly from RandomMixin<FunkyRNG>. This permits
+ * RandomMixin from accessing the derived class's rand64() function, while also allowing
+ * the derived class to provide more.
+ *
+ * The derived class must satisfy the RandomNumberGenerator concept.
  */
-class FastRandomContext
+template<typename T>
+class RandomMixin
 {
 private:
-    bool requires_seed;
-    ChaCha20 rng;
-
     uint64_t bitbuf;
-    int bitbuf_size;
+    int bitbuf_size{0};
 
-    void RandomSeed();
+    /** Access the underlying generator.
+     *
+     * This also enforces the RandomNumberGenerator concept. We cannot declare that in the template
+     * (no template<RandomNumberGenerator T>) because the type isn't fully instantiated yet there.
+     */
+    RandomNumberGenerator auto& Impl() { return static_cast<T&>(*this); }
 
     void FillBitBuffer()
     {
-        bitbuf = rand64();
+        bitbuf = Impl().rand64();
         bitbuf_size = 64;
     }
 
 public:
-    explicit FastRandomContext(bool fDeterministic = false) noexcept;
-
-    /** Initialize with explicit seed (only for testing) */
-    explicit FastRandomContext(const uint256& seed) noexcept;
+    RandomMixin() noexcept = default;
 
-    // Do not permit copying a FastRandomContext (move it, or create a new one to get reseeded).
-    FastRandomContext(const FastRandomContext&) = delete;
-    FastRandomContext(FastRandomContext&&) = delete;
-    FastRandomContext& operator=(const FastRandomContext&) = delete;
+    // Do not permit copying an RNG.
+    RandomMixin(const RandomMixin&) = delete;
+    RandomMixin& operator=(const RandomMixin&) = delete;
 
-    /** Move a FastRandomContext. If the original one is used again, it will be reseeded. */
-    FastRandomContext& operator=(FastRandomContext&& from) noexcept;
+    RandomMixin(RandomMixin&& other) noexcept : bitbuf(other.bitbuf), bitbuf_size(other.bitbuf_size)
+    {
+        other.bitbuf_size = 0;
+    }
 
-    /** Generate a random 64-bit integer. */
-    uint64_t rand64() noexcept
+    RandomMixin& operator=(RandomMixin&& other) noexcept
```

Comment:
While you're at it, how about replacing these move assignments with `Reseed()` functions?

I find the idea of moving an rng to be unintuitive. And as far as I can tell, every case of 

`RandomMixin = foo` and `FastRandomContext = foo` and `InsecureRandomContext = foo` (all in tests) really just means "reseed".